### PR TITLE
Generate controllers and api groups

### PR DESCRIPTION
**Pull request by amy**
Generates controllers and api groups for initial kubeadm provider scaffolding. Fetches kubebuilder for tests.

Closes: https://github.com/kubernetes-sigs/cluster-api-bootstrap-provider-kubeadm/issues/1 and https://github.com/kubernetes-sigs/cluster-api-bootstrap-provider-kubeadm/issues/9

### CODE REVIEW

File: api/v1alpha1/groupversion_info.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+// +groupName=kubeadm.bootstrap.cluster.sigs.k8s.io
+package v1alpha1
+
+import (
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/controller-runtime/pkg/scheme"
+)
+
+var (
+	// GroupVersion is group version used to register these objects
```
Comment (ashish-amarnath) [inline api/v1alpha1/groupversion_info.go]:
```suggestion
	// GroupVersion is group version used to register kubeadm-bootstrap-provider objects
```

File: api/v1alpha1/kubeadmbootstrapconfig_types.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
```
Comment (ashish-amarnath) [inline api/v1alpha1/kubeadmbootstrapconfig_types.go]: suggest removing L23 and L24

File: api/v1alpha1/kubeadmbootstrapconfig_types.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// KubeadmBootstrapConfigSpec defines the desired state of KubeadmBootstrapConfig
+type KubeadmBootstrapConfigSpec struct {
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
```
Comment (ashish-amarnath) [inline api/v1alpha1/kubeadmbootstrapconfig_types.go]:
```suggestion
	// INSERT ADDITIONAL SPEC FIELDS - desired state of KubeadmBootstrapConfig
```

File: api/v1alpha1/kubeadmbootstrapconfig_types.go [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+	// Important: Run "make" to regenerate code after modifying this file
+}
+
+// KubeadmBootstrapConfigStatus defines the observed state of KubeadmBootstrapConfig
+type KubeadmBootstrapConfigStatus struct {
+	// INSERT ADDITIONAL STATUS FIELD - define observed state of cluster
```
Comment (ashish-amarnath) [inline api/v1alpha1/kubeadmbootstrapconfig_types.go]:
```suggestion
	// INSERT ADDITIONAL STATUS FIELD - define observed state of KubeadmBootstrapConfig
```

File: api/v1alpha1/kubeadmbootstrapconfig_types_test.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
```
Comment (ashish-amarnath) [inline api/v1alpha1/kubeadmbootstrapconfig_types_test.go]: Any reason why we aren't using the gloang `"testing"` package instead?

File: config/samples/kubeadm_v1alpha1_kubeadmbootstrapconfig.yaml [hunk 6]
```diff
@@ -0,0 +1,7 @@
+apiVersion: kubeadm.bootstrap.cluster.sigs.k8s.io/v1alpha1
```
Comment (ashish-amarnath) [inline config/samples/kubeadm_v1alpha1_kubeadmbootstrapconfig.yaml]: suggest making this sample more realistic.

File: api/v1alpha1/kubeadmbootstrapconfig_types_test.go (hunk 5 above)
Comment (ashish-amarnath) [reply api/v1alpha1/kubeadmbootstrapconfig_types_test.go]: Throughout

File: go.sum [hunk 7]
```diff
@@ -0,0 +1,123 @@
+cloud.google.com/go v0.26.0/go.mod h1:aQUYkXzVsufM+DwF1aE+0xfcU+56JwCaLick0ClmMTw=
```
Comment (ashish-amarnath) [inline go.sum]: is this file up-to-date?

**Comment by moshloop** [comment]
shouldn't `controllers` and `api` folder be under a `pkg` folder ?

File: api/v1alpha1/kubeadmbootstrapconfig_types_test.go (hunk 5 above)
Comment (moshloop) [reply api/v1alpha1/kubeadmbootstrapconfig_types_test.go]: +2 for `gomega`,  golang `testing` is fine for some basic tests, but once you start doing extensive unit and integration testing that this controller will need the lack of matchers really becomes painful 

+0 for `ginkgo` - I think BDD is an individual style choice, and whether to use it or not probably amounts to how many potential contributors prefer that style.  I personally don't prefer BDD, but I don't dislike it enough to give it a -1

Comment (detiber) [reply api/v1alpha1/kubeadmbootstrapconfig_types_test.go]: This is generated by the kubebuilder scaffolding.

File: go.sum (hunk 7 above)
Comment (chuckha) [reply go.sum]: ./hack/verify-deps.sh will fail if it's not

File: api/v1alpha1/kubeadmbootstrapconfig_types.go (hunk 2 above)
Comment (amy) [reply api/v1alpha1/kubeadmbootstrapconfig_types.go]: This is a generated file and also the beginning of scaffolding. Gonna leave this as a placeholder.

File: api/v1alpha1/kubeadmbootstrapconfig_types.go (hunk 4 above)
Comment (amy) [reply api/v1alpha1/kubeadmbootstrapconfig_types.go]: This is a generated file and also the beginning of scaffolding. Gonna leave this.

File: api/v1alpha1/kubeadmbootstrapconfig_types_test.go (hunk 5 above)
Comment (amy) [reply api/v1alpha1/kubeadmbootstrapconfig_types_test.go]: This is a generated file and also the beginning of scaffolding. Gonna leave this.

**Comment by amy** [comment]
This is all generated code and initial scaffolding. If comments are actually out of scope for this PR, please make a separate issue.

**Comment by vincepri** [comment]
/lgtm

@moshloop Kubebuilder v2 restructures the repository layout and removes the need of pkg/ etc afaik. We can revisit later if needed.

**Comment by moshloop** [comment]
@vincepri no need to revisit if that's the new layout with v2

**Pull request merged**